src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,350 @@
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * An internal utility class that can be used to adapt system-call-styled method handles
+ * for efficient and easy use.
+ */
+public final class CaptureStateUtil {
+
+    private static final StructLayout CAPTURE_LAYOUT = Linker.Option.captureStateLayout();
+    private static final CarrierLocalArenaPools POOL = CarrierLocalArenaPools.create(CAPTURE_LAYOUT);
+
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+
+    // The method handles below are bound to static methods residing in this class
+
+    private static final MethodHandle NON_NEGATIVE_INT_MH =
+            MhUtil.findStatic(LOOKUP, "nonNegative",
+                    MethodType.methodType(boolean.class, int.class));
+
+    private static final MethodHandle SUCCESS_INT_MH =
+            MhUtil.findStatic(LOOKUP, "success",
+                    MethodType.methodType(int.class, int.class, MemorySegment.class));
+
+    private static final MethodHandle ERROR_INT_MH =
+            MhUtil.findStatic(LOOKUP, "error",
+                    MethodType.methodType(int.class, MethodHandle.class, int.class, MemorySegment.class));
+
+    private static final MethodHandle NON_NEGATIVE_LONG_MH =
+            MhUtil.findStatic(LOOKUP, "nonNegative",
+                    MethodType.methodType(boolean.class, long.class));
+
+    private static final MethodHandle SUCCESS_LONG_MH =
+            MhUtil.findStatic(LOOKUP, "success",
+                    MethodType.methodType(long.class, long.class, MemorySegment.class));
+
+    private static final MethodHandle ERROR_LONG_MH =
+            MhUtil.findStatic(LOOKUP, "error",
+                    MethodType.methodType(long.class, MethodHandle.class, long.class, MemorySegment.class));
+
+    private static final MethodHandle ACQUIRE_ARENA_MH =
+            MhUtil.findStatic(LOOKUP, "acquireArena",
+                    MethodType.methodType(Arena.class));
+
+    private static final MethodHandle ALLOCATE_MH =
+            MhUtil.findStatic(LOOKUP, "allocate",
+                    MethodType.methodType(MemorySegment.class, Arena.class));
+
+    private static final MethodHandle ARENA_CLOSE_MH =
+            MhUtil.findVirtual(LOOKUP, Arena.class, "close",
+                    MethodType.methodType(void.class));
+
+    // The `BASIC_HANDLE_CACHE` contains the common "basic handles" that can be reused for
+    // all adapted method handles. Keeping as much as possible reusable reduces the number
+    // of combinators needed to form an adapted method handle.
+    // The map is lazily computed.
+    //
+    private static final Map<BasicKey, MethodHandle> BASIC_HANDLE_CACHE =
+            new ConcurrentHashMap<>();
+
+    // A key that holds both the `returnType` and the `stateName` needed to look up a
+    // specific "basic handle" in the `BASIC_HANDLE_CACHE`.
+    //   returnType E {int.class | long.class}
```

Comment:
I think using `&in;` or `\in` instead of `E` would be more clear.

Reply:
I didn't know about &in; so thanks for this piece of information @liach